File: maxScore.py
'''Explanation:
1. Initialize left_sum and max_sum:
Calculate the sum of the first k elements and store it in left_sum. This represents the initial maximum sum.
2. Iterate from the right:
Iterate from the right side of the array, gradually adding elements to right_sum while subtracting elements from left_sum.
3. Update max_sum:
At each iteration, compare the current left_sum + right_sum with the current max_sum and update max_sum if necessary.
4. Return max_sum:
After the loop, max_sum will contain the maximum points that can be obtained from the cards.'''

class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        left_sum = sum(cardPoints[:k])
        max_sum = left_sum
        right_sum = 0

        for i in range(k - 1, -1, -1):
            left_sum -= cardPoints[i]
            right_sum += cardPoints[len(cardPoints) - k + i]
            max_sum = max(max_sum, left_sum + right_sum)

        return max_sum
        

File: maxScore.ts
function maxScore(cardPoints: number[], k: number): number {
    let n = cardPoints.length; // Get the length of the cardPoints array
    let currentSum = 0; // Initialize currentSum to store the sum of the first k cards

    // Calculate the sum of the first k cards (initial sliding window)
    for (let i = 0; i < k; i++) {
        currentSum += cardPoints[i];
    }

    let maxNum = currentSum; // Initialize maxNum to the sum of the first k cards

    // Use a sliding window to replace cards from the beginning with cards from the end
    for (let i = 0; i < k; i++) {
        // Update currentSum by removing the card from the beginning and adding the card from the end
        currentSum = (currentSum - cardPoints[k - 1 - i]) + cardPoints[n - 1 - i];
        // Update maxNum if the new currentSum is greater
        maxNum = currentSum > maxNum ? currentSum : maxNum;
    }

    return maxNum; // Return the maximum sum of k card
    
};
